From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NeKit <nekit1000@gmail.com>
Date: Sun, 19 Apr 2020 23:47:38 +0200
Subject: [PATCH] (halium) init: ignore non-kernel messages in ReadUevent

Do not abort processing pending uevents in case uevent_kernel_multicast_recv
returns EIO errno.

This helps not to break ueventd coldboot mechanism when running in parallel
with udev which will send its own events to netlink socket.

Change-Id: Ib7e4521899035e2dc67efaf27fc1bea13659efb4
---
 init/uevent_listener.cpp | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/init/uevent_listener.cpp b/init/uevent_listener.cpp
index 7cd396a8b..4d853102f 100644
--- a/init/uevent_listener.cpp
+++ b/init/uevent_listener.cpp
@@ -99,7 +99,19 @@ ReadUeventResult UeventListener::ReadUevent(Uevent* uevent) const {
     char msg[UEVENT_MSG_LEN + 2];
     int n = uevent_kernel_multicast_recv(device_fd_, msg, UEVENT_MSG_LEN);
     if (n <= 0) {
-        if (errno != EAGAIN && errno != EWOULDBLOCK) {
+        if (errno == EIO) {
+            // ignore non-kernel message (probably from udev) but keep processing events
+            uevent->partition_num = -1;
+            uevent->major = -1;
+            uevent->minor = -1;
+            uevent->action.clear();
+            uevent->path.clear();
+            uevent->subsystem.clear();
+            uevent->firmware.clear();
+            uevent->partition_name.clear();
+            uevent->device_name.clear();
+            return ReadUeventResult::kSuccess;
+        } else if (errno != EAGAIN && errno != EWOULDBLOCK) {
             PLOG(ERROR) << "Error reading from Uevent Fd";
         }
         return ReadUeventResult::kFailed;
-- 
2.25.1

